MapboxNavigation/NavigationMapView.swift

```diff
@@ -96,19 +115,113 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
-        let imageView = UIImageView(image: image)
-        imageView.contentMode = .center
-        imageView.backgroundColor = .gray
-        imageView.frame = bounds
-        addSubview(imageView)
+        let contentFrame = UIEdgeInsetsInsetRect(bounds, contentInset)
+        let courseViewWidth = userCourseView?.frame.width ?? 0
+        let courseViewHeight = userCourseView?.frame.height ?? 0
+        let edgePadding = UIEdgeInsets(top: 50 + courseViewHeight / 2,
+                                       left: 50 + courseViewWidth / 2,
+                                       bottom: 50 + courseViewHeight / 2,
+                                       right: 50 + courseViewWidth / 2)
+        return CGPoint(x: max(min(contentFrame.midX,
+                                  contentFrame.maxX - edgePadding.right),
+                              contentFrame.minX + edgePadding.left),
+                       y: max(max(min(contentFrame.minY + contentFrame.height * 0.8,
+                                      contentFrame.maxY - edgePadding.bottom),
+                                  contentFrame.minY + edgePadding.top),
+                              contentFrame.minY + contentFrame.height * 0.5))
+    }
+    
+    /**
+     Determines whether the map should follow the user location and rotate when the course changes.
+     - seealso: NavigationMapViewCourseTrackingDelegate
+     */
+    open var tracksUserCourse: Bool = false {
+        didSet {
+            if tracksUserCourse {
+                enableFrameByFrameCourseViewTracking(for: 2)
+                altitude = NavigationMapView.defaultAltitude
+                showsUserLocation = true
+                courseTrackingDelegate?.navigationMapViewDidStartTrackingCourse?(self)
+            } else {
+                courseTrackingDelegate?.navigationMapViewDidStopTrackingCourse?(self)
+            }
+            
+            if let location = userLocationForCourseTracking {
+                updateCourseTracking(location: location, animated: true)
+            }
+        }
+    }
+
+    /**
+     A `UIView` used to indicate the user’s location and course on the map.
+     
+     If the view conforms to `UserCourseView`, its `UserCourseView.update(location:pitch:direction:animated:)` method is frequently called to ensure that its visual appearance matches the map’s camera.
+     */
+    @objc public var userCourseView: UIView? {
+        didSet {
+            oldValue?.removeFromSuperview()
+            if let userCourseView = userCourseView {
+                if let location = userLocationForCourseTracking {
+                    updateCourseTracking(location: location, animated: false)
+                } else {
+                    userCourseView.center = userAnchorPoint
+                }
+                addSubview(userCourseView)
+            }
+        }
     }
     
+    private lazy var mapTapGesture = UITapGestureRecognizer(target: self, action: #selector(didRecieveTap(sender:)))
```

Comment:
Curious, why are we using `UITapGestureRecognizer` vs `Touches` for tapping our mapView? 
> Gesture recognizers may delay the delivery of touch objects to the view while they are recognizing gestures, and by default they cancel delivery of remaining touch objects to the view once they recognize their gesture.
[Reference: UIGestureRecognizers vs touches (accuracy)
](https://stackoverflow.com/a/11851706)

Reply:
PS: I know it's existing code, but this is just a general question for everyone. 
cc @mapbox/navigation-ios

Reply:
The map SDK implements its built-in gestures using gesture recognizers rather than interpreting touches manually, so NavigationMapView would need to use gesture recognizers to avoid conflicts.

The UIGestureRecognizer API offers several ways to resolve conflicts between gesture recognizers. If you encounter a case where the map SDK exhibits improper gesture behavior, please open a ticket upstream in the mapbox-gl-native repository. Thanks!

Reply:
So, really if the map SDK uses gesture recognizers, then we are forced to use gesture recognizers because using any of the touches delegates will get canceled. Which makes sense.